src/charm.py

```diff
@@ -55,8 +44,16 @@ class KyuubiCharm(ops.CharmBase):
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.workload = KyuubiServer(self.unit.get_container(KYUUBI_CONTAINER_NAME))
-        self.s3_requirer = S3Requirer(self, S3_INTEGRATOR_REL)
+
+        # Workload
+        self.workload = KyuubiWorkload(
+            container=self.unit.get_container(KYUUBI_CONTAINER_NAME),
+        )
+
+        # Context
+        self.context = Context(self)
+
+        # Requirers
         self.metastore_db = DatabaseRequires(
```

Comment:
I have moved the `requires` objects to the main charm class, because in the current implementation, the `context` object holds all the state information (regarding metastore db information, auth db information ,s3 information and service account information). Reading the relation data from `Context` class is as straightforward for database relation as it is for s3 relation, because for database relation, the username and password are stored in the secrets. To view them in plaintext, `db_requires.fetch_relation_data()` is used. However, it is not straightforward for the `Context` class to access `db_requires` from inside the event handlers (I feel like it is anti pattern). That's why, I moved the requires relation definition to the charm class so that everyone interested can access it from there.

Reply:
And for consistency, I have moved S3 requires and Provider relation definition to the main class as well.

Reply:
This is actually my main concern. This does not quite follow the pattern. The idea here is that we should not use the `DatabaseRequires` anymore, but swap to use the `*Data` classes and `*EventsHandlers` classes. 

The `Data` classes should in general be wrapped by the context, whereas the events handlers should be imported and used in the events classes. 

See [here](https://github.com/canonical/spark-integration-hub-k8s-operator/blob/main/src/core/context.py#L25) and [here](https://github.com/canonical/karapace-operator/blob/main/src/core/cluster.py#L42) for the context. And [here](https://github.com/canonical/karapace-operator/blob/main/src/events/kafka.py#L32) for the events handlers.

Reply:
I agree with this one. I'll refactor this so that we use `*Data` and `*EventHandler` classes separately.